lib/pool.dart

```diff
@@ -121,6 +121,44 @@ class Pool {
     });
   }
 
+  /// Requests a resource for the combined duration of the [callback] and the
+  /// processing of the Stream it returns.
+  /// (E.g. a database connection is hold until the the results get processed.)
+  Stream/*<T>*/ streamWithResource/*<T>*/(Stream/*<T>*/ callback()) {
+    StreamController/*<T>*/ streamController;
+    StreamSubscription/*<T>*/ streamSubscription;
+    Completer resourceCompleter = new Completer();
+    final complete = () {
+      if (streamSubscription != null) {
+        streamSubscription.cancel();
+        streamSubscription = null;
+      }
+      if (!resourceCompleter.isCompleted) {
+        resourceCompleter.complete();
+        streamController.close();
+      }
+    };
+    streamController = new StreamController/*<T>*/(onCancel: complete);
+    withResource(() {
+      if (resourceCompleter.isCompleted) return null;
+      new Future.sync(callback).then((Stream/*<T>*/ stream) {
+        if (stream == null) {
+          complete();
+          return null;
+        }
+        streamSubscription = stream.listen(streamController.add,
+            onError: streamController.addError,
+            onDone: complete,
+            cancelOnError: true);
```

Comment:
> Not sure how this should work. The returned Future of `withResource` will complete, it doesn't matter if we add an `await` or `.then()` on it. I think the same should be true for Streams (e.g. it should complete with or without a subscriber), but can be convinced otherwise.

This is one of the differences between Futures and Streams in general: Futures don't care about whether a listener is registered and streams do (or at least, they can). Given that, I think it makes sense for the behavior to differ here. In practice, I don't expect anyone will realistically be passing streams through only to discard them later. It's more likely that they'll want to make use of the laziness and memory benefits that come from following stream conventions here.

> sure, thanks for jumping on this!

I'll cc you on the pull request.

> How can I listen on Stream's done? The `drain` will introduce another listener, which will fail if it is single-subscription. `StreamController.done` is not available when all I got is a Stream.

You can always use a transformer to hook into a stream's events without consuming it, just like you're doing in this PR already.